Sources/GRPC/AsyncAwaitSupport/GRPCClient+AsyncAwaitSupport.swift

```diff
@@ -148,4 +148,314 @@ extension GRPCClient {
+  }
+
+  public func performAsyncBidirectionalStreamingCall<
+    Request: GRPCPayload,
+    Response: GRPCPayload,
+    RequestStream: Sequence
+  >(
+    path: String,
+    requests: RequestStream,
+    callOptions: CallOptions? = nil,
+    interceptors: [ClientInterceptor<Request, Response>] = [],
+    requestType: Request.Type = Request.self,
+    responseType: Response.Type = Response.self
+  ) -> GRPCAsyncResponseStream<Response>
+    where RequestStream.Element == Request {
+    let call = self.channel.makeAsyncBidirectionalStreamingCall(
+      path: path,
+      callOptions: callOptions ?? self.defaultCallOptions,
+      interceptors: interceptors
+    )
+    return self.perform(call, with: AsyncStream(wrapping: requests))
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension GRPCClient {
+  @inlinable
+  internal func perform<Request, Response, RequestStream>(
+    _ call: GRPCAsyncClientStreamingCall<Request, Response>,
+    with requests: RequestStream
+  ) async throws -> Response
+    where RequestStream: AsyncSequence, RequestStream.Element == Request {
+    // We use a detached task because we use cancellation to signal early, but successful exit.
+    let requestsTask = Task.detached {
+      try Task.checkCancellation()
+      for try await request in requests {
+        try Task.checkCancellation()
+        try await call.sendMessage(request)
+      }
+      try Task.checkCancellation()
+      try await call.sendEnd()
+      try Task.checkCancellation()
+    }
+    return try await withTaskCancellationHandler {
+      // Await the response, which may come before the request stream is exhausted.
+      let response = try await call.response
+      // If we have a response, we can stop sending requests.
+      requestsTask.cancel()
+      // Return the response.
+      return response
+    } onCancel: {
+      requestsTask.cancel()
+      // If this outer task is cancelled then we should also cancel the RPC.
+      Task.detached {
+        try await call.cancel()
+      }
+    }
+  }
+
+  @inlinable
+  internal func perform<Request, Response, RequestStream>(
+    _ call: GRPCAsyncBidirectionalStreamingCall<Request, Response>,
+    with requests: RequestStream
+  )
+    -> GRPCAsyncResponseStream<Response>
+    where RequestStream: AsyncSequence, RequestStream.Element == Request {
+    Task {
+      try await withTaskCancellationHandler {
+        try Task.checkCancellation()
+        for try await request in requests {
+          try Task.checkCancellation()
+          try await call.sendMessage(request)
+        }
+        try Task.checkCancellation()
+        try await call.sendEnd()
+      } onCancel: {
+        Task.detached {
+          try await call.cancel()
+        }
+      }
+    }
+    return call.responses
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension AsyncStream {
+  fileprivate init<T>(wrapping sequence: T) where T: Sequence, T.Element == Element {
```

Comment:
Mind adding a note here that we could special case the request stream being a `Sequence` to avoid going via `AsyncStream`?